### PR TITLE
Test ethrelay catchup and test challenge invalid block

### PR DESCRIPTION
**Pull request by ailisp**
Implement #238

### CODE REVIEW

File: environment/lib/near2eth-relay.js [hunk 1]
```diff
@@ -170,6 +170,100 @@ class Near2EthRelay {
...
+                console.log(`Encountered error while requesting light client block ${err}`);
+                await sleep(1000);
+            }
+        }
+        console.log(`${JSON.stringify(lightClientBlock)}`);
+        while (true) {
+            let borshBlock = borshify(lightClientBlock);
+            console.log('Mutate block by one byte')
+            console.log(borshBlock)
+            borshBlock[Math.floor(borshBlock.length * Math.random())] += 1
```
Comment (nearmax) [inline environment/lib/near2eth-relay.js:245]: What if byte is already 255? Will it wrap around or saturate?

File: ci/test_challenge.sh [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+sleep 5
+yarn run pm2 list
+node index.js start eth-relay
+sleep 5
+yarn run pm2 list
+node index.js start near-watchdog
+sleep 5
+yarn run pm2 list
+
+sleep 30
```
Comment (nearmax) [inline ci/test_challenge.sh:50]: Awesome, I see you had to increase the bond time to reduce flakiness.

File: environment/index.js [hunk 3]
```diff
@@ -443,6 +444,24 @@ RainbowConfig.addOptions(
     ],
 );
 
+// Testing command
+const dangerCommand = program.command('DANGER');
```
Comment (nearmax) [inline environment/index.js:448]: Could we add description to DANGER command and submit_invalid_near_block command?

File: environment/lib/near2eth-relay.js (hunk 1 above)
Comment (ailisp) [reply environment/lib/near2eth-relay.js:245]: It wrap around:
```
> a=Buffer([255])
<Buffer ff>
> a[0]+=1
256
> a
<Buffer 00>
```

File: ci/test_challenge.sh (hunk 2 above)
Comment (ailisp) [reply ci/test_challenge.sh:50]: Yeah, 10s lock time was too short to send the challenge

**Comment by ailisp** [comment]
@nearmax got this error in e2e test chanllenge, this's also always happens locally (Seems there's either a problem with watchdog, or contract)
```
[2020-08-07T22:36:14Z] (node:5160) UnhandledPromiseRejectionWarning: Error: [-32000] Server error: 5GkUH4hzeZs17nrL8wSrsNLt7ZHJ8ApgXxrtRk6pD79v does not exist
...
[2020-08-07T22:36:14Z]     at async Command.execute (/var/lib/buildkite-agent/builds/buildkite-i-01dd6e880ebe52ccf-1/nearprotocol/rainbow-bridge/environment/commands/transfer-eth-erc20-from-near.js:148:24)
```
Underlying code is https://github.com/near/rainbow-bridge/blob/master/environment/commands/transfer-eth-erc20-from-near.js#L148

Do you have idea why a light_client_proof respond "does not exist" on requested receiptId? The receiptId is responded by rpc of the successful burn txn so it should exist. Also @bowenwang1996

**Comment by nearmax** [comment]
> @nearmax got this error in e2e test chanllenge, this's also always happens locally (Seems there's either a problem with watchdog, or contract)
> 
> ```
> [2020-08-07T22:36:14Z] (node:5160) UnhandledPromiseRejectionWarning: Error: [-32000] Server error: 5GkUH4hzeZs17nrL8wSrsNLt7ZHJ8ApgXxrtRk6pD79v does not exist
> ...
> [2020-08-07T22:36:14Z]     at async Command.execute (/var/lib/buildkite-agent/builds/buildkite-i-01dd6e880ebe52ccf-1/nearprotocol/rainbow-bridge/environment/commands/transfer-eth-erc20-from-near.js:148:24)
> ```
> 
> Underlying code is https://github.com/near/rainbow-bridge/blob/master/environment/commands/transfer-eth-erc20-from-near.js#L148
> 
> Do you have idea why a light_client_proof respond "does not exist" on requested receiptId? The receiptId is responded by rpc of the successful burn txn so it should exist. Also @bowenwang1996

Is it 100% reproducible? Or does it happen randomly? There might be a race condition issue when ViewClient hasn't learnt yet that this receipt exists. I've seen similar issue with light client header -- RPC would tell me that this header exists, but if I try to immediately grab it with light client endpoint it would tell me it is not know, but then if I wait it becomes known.

**Comment by bowenwang1996** [comment]
> Do you have idea why a light_client_proof respond "does not exist" on requested receiptId? The receiptId is responded by rpc of the successful burn txn so it should exist. Also @bowenwang1996

You need to wait for the next block for the outcome root.

**Comment by ailisp** [comment]
> You need to wait for the next block for the outcome root.

The next block you mean "near block" right? I waited for a few minutes and retry the rpc manually and it still "does not exist".

**Comment by ailisp** [comment]
> Is it 100% reproducible? Or does it happen randomly? There might be a race condition issue when ViewClient hasn't learnt yet that this receipt exists. I've seen similar issue with light client header -- RPC would tell me that this header exists, but if I try to immediately grab it with light client endpoint it would tell me it is not know, but then if I wait it becomes known.

Seems always reproducible locally, and wait doesn't help (tried print the params and retry rpc a few minutes later and just now (10+ hours later), still not exist.

**Comment by bowenwang1996** [comment]
What does the set up look like? How many near nodes are there and what is the epoch length?

**Comment by ailisp** [comment]
> What does the set up look like? How many near nodes are there and what is the epoch length?

It's default near init setup, one node, epoch length is 60

**Comment by bowenwang1996** [comment]
I wonder whether it is because block production is too fast and garbage collection kicks in very quickly. Does it help if you change epoch length to 600 or make the node archival?

**Comment by ailisp** [comment]
> I wonder whether it is because block production is too fast and garbage collection kicks in very quickly. Does it help if you change epoch length to 600 or make the node archival?

Good point!

**Comment by nearmax** [comment]
> > I wonder whether it is because block production is too fast and garbage collection kicks in very quickly. Does it help if you change epoch length to 600 or make the node archival?
> 
> Good point!
> Update: it's probably the reason! passed first time, the restart with new config.json couldn't automated yet, but it passes. At the meantime, i've log and manually call rpc that verified every hash calculation is correct in NearBridge.sol. So we're good now. Thanks for help!

I've seen this issue before with bridge e2e tests, too bad it did not occur to me that this is the case. Thank you @bowenwang1996 !

**Pull request merged**